fix dotnet run infinite loop by rendering help explicitly

baronfel opened the pull request:
Fixes #23687 and #23034 by directly returning a HelpContext InvocationResult in the help middleware pipeline, instead of trying to force the system to kinda-sorta invoke the HelpContext itself on a manually-altered parameter string.

baronfel committed:
fix dotnet run infinite loop by rendering help explicitly

baronfel commented on src/Cli/dotnet/Parser.cs:
```diff
@@ -118,6 +118,19 @@ public static Command GetBuiltInCommand(string commandName)
             .EnableLegacyDoubleDashBehavior()
             .Build();
 
+         class HelpResult : IInvocationResult
```
This HelpResult is a copy/paste of the HelpResult type in the command-line-api repo [here](https://github.com/dotnet/command-line-api/blob/main/src/System.CommandLine/Help/HelpResult.cs). This type isn't internal so it must be reproduced in full here.

dsplaisted commented on src/Cli/dotnet/Parser.cs:
```diff
@@ -118,6 +118,19 @@ public static Command GetBuiltInCommand(string commandName)
+                TextWriter output = context.Console.Out.CreateTextWriter();
+                var helpBuilder = (HelpBuilder)context.BindingContext.GetService(typeof(HelpBuilder));
+                var console = context.Console as SystemConsole;
+                var width = (console?.IsOutputRedirected ?? false) ? int.MaxValue : Console.WindowWidth;
```
It seems like `width` isn't used, and that this line is dead code.

dsplaisted reviewed (approved):
Can we add automated test coverage for this?

This seems like a regression in 6.0.200, do you know how that happened since the command line parsing code was supposed to be reverted to what we had in 6.0.100?

Shouldn't we submit this for approval for 6.0.200?  It seems like a somewhat bad regression.  /cc @marcpopMSFT

baronfel committed:
remove HandleMissingCommand explicit command handler in favor of using middleware

this was done because the dependencies required to construct the HelpContext are only resolvable from the InvocationContext

baronfel committed:
remove HelpException

It's not used anymore and we have more standard ways of triggering help

baronfel committed:
treat unmatchedtokensaserrors set to false for commands that forward to msbuild and other tools

baronfel commented:
This has grown somewhat as I look at fixing callsites that required the erroring ShowHelp method:

* `ShowHelp` removed in favor of explicit HelpResult in the error handling middleware, requires
    * `HandleMissingCommand` removed in favor of adding a new middleware that detects the 'no handler configured for this command' case and writes the previous error message + triggers help, requires
        * setting `TreatUnknownTokensAsError = false` for commands that forward parameters to prevent parse errors
    * other cases where `ShowHelp()` was triggered explicitly instead happen as part of the Exception handling middleware now

This seems to be working well with passthrough parameters in my local testing, but the size of the PR has grown somewhat.

baronfel committed:
remove unused width

baronfel commented on src/Cli/dotnet/Program.cs:
```diff
@@ -64,33 +64,6 @@ public static int Main(string[] args)
                 {
                     return ProcessArgs(args, startupTime);
                 }
-                catch (HelpException e)
```
the use of ShowHelp() here means that to get the same semantics we needed to push these clauses to the Exception handling middleware in the Parser definition.

baronfel commented on src/Cli/dotnet/commands/dotnet-add/AddCommandParser.cs:
```diff
@@ -33,8 +33,6 @@ private static Command ConstructCommand()
             command.AddCommand(AddPackageParser.GetCommand());
             command.AddCommand(AddProjectToProjectReferenceParser.GetCommand());
 
-            command.SetHandler((parseResult) => parseResult.HandleMissingCommand());
```
I removed all use of HandleMissingCommand in favor of a middleware that shows help when a command is triggered and that command doesn't have an explicit command.

baronfel commented on src/Cli/dotnet/commands/dotnet-build/BuildCommandParser.cs:
```diff
@@ -76,6 +76,8 @@ private static Command ConstructCommand()
             command.AddOption(CommonOptions.ArchitectureOption);
             command.AddOption(CommonOptions.OperatingSystemOption);
 
+            command.TreatUnmatchedTokensAsErrors = false;
```
I went through each command, and anytime I found commands that needed to accept arbitrary tokens to forward to another binary, I added this flag.

baronfel commented on src/Cli/dotnet/commands/dotnet-nuget/NuGetCommandParser.cs:
```diff
@@ -63,7 +63,7 @@ private static Command GetLocalsCommand()
 
             localsCommand.AddArgument(new Argument<string>()
-                .FromAmong(new string[] { "all", "http-cache", "global-packages", "plugins-cache", "temp" }));
+                .FromAmong(new string[] { "all", "http-cache", "global-packages", "temp" }));
```
I invoked the raw nuget CLI via `dotnet <path to nuget CLI dll> --help` and found that the list of known values for this argument in the SDK help didn't match the help from nuget directly.

baronfel commented on src/Cli/dotnet/commands/dotnet-nuget/NuGetCommandParser.cs:
```diff
@@ -122,7 +122,7 @@ private static Command GetTrustCommand()
                          .FromAmong(new string[] { "list", "author", "repository", "source", "certificate", "remove", "sync" }));
 
-            trustCommand.AddOption(new Option<string>("--algorithm"));
+            trustCommand.AddOption(new Option<string>("--algorithm").FromAmong("SHA256", "SHA384", "SHA512"));
```
same here, the algorithms listed in nuget help are a thing that we can forward along.

it would be nice to consider asking the NuGet team for a System.Commandline Command that we could seamlessly integrate and not have to manually patch over like this.  @clairernovotny would this be a good thing to track?

zivkan replied:
This issue is basically that: https://github.com/NuGet/Home/issues/10039

Problem from NuGet.Client's side is how to prioritize.

baronfel replied:
Thank you for the link, @zivkan! Re: prioritization, I think it's a nice to have. If the NuGet team wants better/complete/up-to-date tab-suggestion that would be a good way to get it, but the current mechanism here seems to be mostly working.  Perhaps worth a maintenance task every so often to check the help output?

baronfel committed:
attempt to remove double-dash legacy usage

baronfel commented:
Closing in favor of the simpler #24449.
